File: new/two_sum_II_two_pointers.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        // idea: use two pointer
        // track left and right numbers
        // if they add up to target, return their indices (+ 1)
        // if their sum < target, increase left pointer, if their sum > target, decrease right pointer
        // do this until left < right
        int left = 0;
        int right = numbers.size() - 1;
        while (left < right) {
            int val = numbers[left] + numbers[right];
            if (val < target) {
                left++;
            } else if (val > target) {
                right--;
            } else {
                return {left + 1, right + 1};
            }
        }
        return {};
    }
    // time: O(n)
    // space: O(1)
};

File: new/python/two_sum_II_two_pointers.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        # idea: two pointers
        # track the left and right number
        # if the add up to target, return their index
        # if the sum is over the target, reduce the right pointer down by 1
        # if the sum is less than the target, increase the left pointer by 1
        # do this until left < right
        left = 0
        right = len(numbers) - 1
        while left < right:
            value = numbers[left] + numbers[right]
            if value == target:
                return [left + 1, right + 1]
            elif value < target:
                left += 1
            else:
                right -= 1
        return []
        # time: O(n)
        # space: O(1)
